Zmq pr

jmxpearson opened the pull request:
Based on @rwschonberg's #184, but with substantial simplifications. 

- Removed lots of code in `nexus.py`, including removal of all special-casing for GUI and TUI.
- Distilled a single quit path in Nexus that simplifies exit logic.
- Rewrote `demos/bubblewrap` to use new 0MQ actors. This establishes a new pattern for future GUIs to adopt.
- Introduced some additional error handling to reduce brittleness and user gotchas.

Commit:
Implement IPC with ZMQ

jmxpearson committed:
check whether GUI process exists before killing

jmxpearson committed:
revamping bubblewrap demo

- reformatting yaml file for new links
- changing readme for new formats

jmxpearson committed:
WIP: revamping GUI pattern

- rewriting GUI as a single actor with an encapsulated state that's
  passed to QT front end
- trying to remove special-case GUI code (commented out in nexus.py)
- currently not working; starting GUI seems to leave zmq sockets in a
  bad state; calling exec function may block

jmxpearson committed:
more tweaks

jmxpearson committed:
most things working now

- quitting GUI sends quit but it doesn't work
- need to overhaul comms with nexus to remove special cases

jmxpearson committed:
adjusting fixtures for new actor config

- currently need logging to be set up before registering with nexus
- probably want to have a default that works without this in future
  (perhaps with a warning)

jmxpearson committed:
removing tests for GUI special-casing

jmxpearson committed:
fixing link test

jmxpearson committed:
renaming messages from nexus to actors

- these messages now start with Nexus, so that all message types start
  with their _source_

jmxpearson committed:
adding new message types

- Actor Signal messages originate at actors as control signals back to
  nexus

jmxpearson committed:
removing unused testing code

jmxpearson committed:
using actorsignalmsg instead of state message

jmxpearson committed:
removing unused code

jmxpearson committed:
change start_nexus to no inputs

- this was parameterized in order to test some specific functionality,
  but that led to much harder to read code that was only leveraged in a
  single test case

jmxpearson committed:
spelling error

jmxpearson committed:
set actor port to control port

- on the way to removing it as an option

jmxpearson committed:
streamlining nexus logic

- now there's a single flow for processing actor input, branching based
  on message type
- still work to do on simplifying quit pathway (better now!) and
  removing dead code
- still some problems with TUI signaling via the output port

jmxpearson committed:
fixing nexus/tui communication

- got rid of extra async that was slowing down quitting
- TUI now quits if it gets QUIT in reply on nexus control port also
  (probably good to leave in quit option on other port, too)
- direct send of text to TUI doesn't seem to be working, though
- TUI logging needs to know log pull port (not available at spinup)

jmxpearson committed:
improved tui/nexus comms

- quit now works well
- tui sends "ready" on startup and prints "Awaiting input:"
- "QUIT" is now broadcast from quit path, not reply on signal port

jmxpearson committed:
logging tweaks

jmxpearson committed:
making sure tui mocks work

jmxpearson committed:
testing mode for TUI app

- this makes sure TUI doesn't try to "phone home" to Nexus and send a
  ready signal during app testing

jmxpearson committed:
fixing error in test_nexus

jmxpearson committed:
updates to enable public log input port

- previously, we only published logging sub port; actors couldn't post
  directly to logger without being spawned from nexus
- now we pick pub port at runtime from cli; this allows, e.g., TUI and
  GUI(s) access to logger as well
- lots of updates to tests

jmxpearson committed:
WIP: currently broken

jmxpearson committed:
tests passing, demos working

- still need to verify TUI logging works, but nothing is throwing an
  error
- still don't like cli.py names for logging ports, but pub and pull are
  reasonable internal names

jmxpearson committed:
TUI logging up and running

- make sure you sanitize your ports correctly!

jmxpearson committed:
working on bubblewrap demo

- comms between GUI and Nexus appear to be working
- current troubles are with reading data and processing actor errors

jmxpearson committed:
fixing bubblewrap errors

- need to guard against TimeoutError when we try to get data
- fixing some other bugs from previous ways of doing things

jmxpearson committed:
more WIP getting bubblewrap demo working

- need to find a way for spawned actors to get a store interface

jmxpearson committed:
move store interface creation to actors

- this allows it to work with both fork and spawn
- removed some dead code

jmxpearson committed:
cli now asks to keep waiting if ports not read

- test strategy for this is tricky, so no test at present

jmxpearson committed:
making sure nexus responds to unknown actor msg

- issues a warning and ignores
- removed some dead code

jmxpearson committed:
refactoring actor signaling

- now all signal handlers share a more fault-tolerating means of
  communicating from nexus to actors
- can make this safer by having retries, but at least now it errors
  correctly

jmxpearson committed:
fixed bug

- make sure actor signal sockets are set up in setup

jmxpearson committed:
failure to call setup_logging now raises exception

- at least reminds the user they forgot to call it

jmxpearson committed:
adding send command to link

- send combines put and get so that REQ/REP sockets are left in a
  consistent state
- putting without getting the response is a huge footgun for those using
  the class to write their own actors

jmxpearson committed:
adding docs for logging files

- docs break down various log files and what info gets filtered into
  them

jmxpearson committed:
formatting changes

jmxpearson committed:
formatting tests

jmxpearson committed:
formatting and linting

jmxpearson committed:
fixing formatting

jmxpearson committed:
more format fixes

draelos commented on demos/bubblewrap/bubble_demo.yaml:
I think sources/sinks and the lots of extra lines make it harder to parse. Could fix it subsequent PR but even a list of input/output connections might be nicer. Also previously direction didn't technically matter -- does it now?

jmxpearson replied:
Directionality should matter. A listens to B shouldn't necessarily imply B listens to A. Since the default is now who subscribes to what ZMQ messages, I think it's a distinction worth keeping. I suggest we push any cosmetic overhauls to a future PR in the interest of wrapping this up.

draelos commented on demos/bubblewrap/actors/front_end.py:
We should have this actor (inside visual.py?) also listen for signals; typing run in the command line doesn't update the visuals unless the button run on the front_end is selected.

jmxpearson replied:
Fixed. I hadn't overloaded `run_step` in `visual.py`.

draelos commented on demos/minimal/actors/sample_spawn_processor.py:
Got an error here Processor has no attribute _get_store_interface

Also -- error only written to global.log, not improv-debug.log. What are the rules here?

jmxpearson replied:
Fixed. We are now agnostic to fork vs. spawn except as determined in the yaml file, so we don't need this actor any more. Will write docs for the various logs and what goes where.

draelos commented on improv/store.py:
Based on current benchmarking, in many cases plasma is faster than Redis. Can we keep legacy code in another file and still enable this use case?

jmxpearson replied:
I'd propose two things:
1. Remove the plasma dependency, with instructions about how to install optionally.
2. We need to have a `PlasmaStoreInterface` that matches the new API. At a glance, the old code here seems fine, but it would be untested code. My vote would be to put it in a separate file in a folder that doesn't get scanned by `pytest` (a demo?) so it doesn't show up in test coverage reports.
 
I think it's a quick PR for someone to do, and I have someone who could probably get docs and class done quickly (or you may)?

jmxpearson replied:
We'd also need to modify `create_store_interface` in Nexus to switch class based on a flag in the yaml. All the more reason to have a demo, IMHO.

draelos reviewed (changes requested):
All tests passed. Minor comments posted on relevant files.

jmxpearson committed:
bubblewrap actor searches for file given relpath

- new load_data function searches `sys.path` for a file given only
  relative file path

jmxpearson committed:
fixing typo

jmxpearson committed:
run in TUI actually runs GUI

needed to overload runStep to call update on the GUI

jmxpearson committed:
getting rid of dead code

jmxpearson committed:
rename runStep -> run_step

this comports with new API in working demos

jmxpearson committed:
adding data logging back into sample_processor

jmxpearson committed:
removing spawn actor in minimal demo

we now handle fork and spawn similarly enough that we don't need a
special actor here any more

jmxpearson committed:
make tui tests wait longer

jmxpearson committed:
waiting even longer

jmxpearson committed:
longer wait for store to die

jmxpearson committed:
just try tui tests

jmxpearson committed:
even longer store waiting

jmxpearson committed:
attempting to guard against TUI hangs

jmxpearson committed:
reraising canceled task errors

jmxpearson committed:
reducing tests

jmxpearson committed:
linger longer

jmxpearson committed:
try not closing log handler

jmxpearson committed:
removing dead line

jmxpearson committed:
slowing down generator

draelos committed:
Added line for logging allowing run; existing line left (516) but never run on my end.

draelos committed:
Updated scipy.signal.windows import as of scipy v1.13.

draelos committed:
Silence millions of jax info messages

draelos committed:
Remove unused spawn variant in bubble demo.

draelos committed:
Remove depricated .py way of launching demos in improv

draelos reviewed (approved):
Tweaked a few minor things but everything is essentially addressed from earlier requested changes, besides things we will address in future PRs.
